File: pubnub-java/src/main/java/com/pubnub/api/endpoints/objects_api/utils/PNSortKey.java
package com.pubnub.api.endpoints.objects_api.utils;

import lombok.AccessLevel;
import lombok.Getter;

import static com.pubnub.api.endpoints.objects_api.utils.PNSortKey.Dir.ASC;
import static com.pubnub.api.endpoints.objects_api.utils.PNSortKey.Dir.DESC;

final public class PNSortKey {
    public enum Dir {
        ASC("asc"), DESC("desc");

        @Getter(AccessLevel.PACKAGE)
        private final String dir;

        Dir(String dir) {
            this.dir = dir;
        }
    }

    public enum Key {
        ID("id"), NAME("name"), UPDATED("updated");

        @Getter(AccessLevel.PACKAGE)
        private final String fieldName;

        Key(String fieldName) {
            this.fieldName = fieldName;
        }
    }

    @Getter
    private final Dir dir;

    @Getter
    private final Key key;

    private PNSortKey(Key key, Dir dir) {
        this.key = key;
        this.dir = dir;
    }

    public static PNSortKey of(Key key, Dir dir) {
        return new PNSortKey(key, dir);
    }

    public static PNSortKey of(Key key) {
        return new PNSortKey(key, ASC);
    }

    public static PNSortKey asc(Key key) {
        return new PNSortKey(key, ASC);
    }

    public static PNSortKey desc(Key key) {
        return new PNSortKey(key, DESC);
    }

    public String toSortParameter() {
        return key.fieldName + ":" + dir.dir;
    }

}

File: pubnub-kotlin/pubnub-kotlin-api/src/commonMain/kotlin/com/pubnub/api/models/consumer/objects/PNSortKey.kt
package com.pubnub.api.models.consumer.objects

sealed interface SortField {
    val fieldName: String
}

enum class PNKey(override val fieldName: String) : SortField {
    ID("id"),
    NAME("name"),
    UPDATED("updated"),
    TYPE("type"),
    STATUS("status"),
}

enum class PNMembershipKey(override val fieldName: String) : SortField {
    CHANNEL_ID("channel.id"),
    CHANNEL_NAME("channel.name"),
    CHANNEL_UPDATED("channel.updated"),
    CHANNEL_STATUS("channel.status"),
    CHANNEL_TYPE("channel.type"),
    UPDATED("updated"),
    STATUS("status"),
    TYPE("type"),
}

enum class PNMemberKey(override val fieldName: String) : SortField {
    UUID_ID("uuid.id"),
    UUID_NAME("uuid.name"),
    UUID_UPDATED("uuid.updated"),
    UUID_STATUS("uuid.status"),
    UUID_TYPE("uuid.type"),
    UPDATED("updated"),
    STATUS("status"),
    TYPE("type"),
}

sealed class PNSortKey<T : SortField>(
    internal val key: T,
    internal val dir: String = "asc",
) {
    class PNAsc<T : SortField>(key: T) : PNSortKey<T>(key = key, dir = "asc")

    class PNDesc<T : SortField>(key: T) : PNSortKey<T>(key = key, dir = "desc")

    fun toSortParameter(): String {
        return key.fieldName + ":" + dir
    }

    companion object {
        fun asc(key: PNKey): PNSortKey<PNKey> {
            return PNAsc(key)
        }

        fun desc(key: PNKey): PNSortKey<PNKey> {
            return PNDesc(key)
        }
    }
}
